Add library with function to spawn a process connected to a pseudo-terminal

Rinzwind opened the pull request:
This pull request adds a small library with a function to spawn a process connected to a pseudo-terminal. As discussed in [pavel-krivanek/terminal issue #2](https://github.com/pavel-krivanek/terminal/issues/2) and [PTerm issue #58](https://github.com/lxsang/PTerm/issues/58), it’s not entirely possible to replicate this function in-image using `posix_spawn`. The function calls `setsid` and `ioctl` with `TIOCSCTTY` in the spawned process. For the `setsid` call, the attribute flag `POSIX_SPAWN_SETSID` can be passed to `posix_spawn` (nonstandard but supported by the GNU C Library and macOS). But for the `ioctl` call there is, as far as I know, no corresponding attribute or file action that can be passed.

The new File Browser that is being adopted (see: [Pharo pull request #16084](https://github.com/pharo-project/pharo/pull/16084)) offers a context menu item ‘Open terminal here’. That currently does not allow using PTerm ([PTerm issue #62](https://github.com/lxsang/PTerm/issues/62)), but it would be nice if it did and if the VM made that convenient to use by providing the necessary supporting library.

Rinzwind committed:
Added library with a function to spawn a process connected to a pseudo-terminal.

guillep commented:
Hi @Rinzwind, @pavel-krivanek nice!

The build had a hiccup, I relaunched it.
I see no objections to this.

It would be nice however to see some comments in the code.
Maybe copy-pasting part of the PR description is enough?

Rinzwind committed:
Added comment to ‘tty_spawn’ describing its functionality and the motivation for providing it through a library included with the VM.

Rinzwind committed:
Added definition for ‘_XOPEN_SOURCE’ to ‘tty.c’.

Rinzwind committed:
Merged commit 55f077d787401be7 (latest ‘pharo-10’).

Rinzwind committed:
Made ‘tty_spawn’ also check for errors when opening the slave pseudo-terminal device and setting up the standard file descriptors.

Rinzwind committed:
Changed exit status for errors in the setup of the child process in ‘tty_spawn’ from 1 to 127 (similar to ‘posix_spawn’).

Rinzwind commented:
I added a comment, fixed the compilation error on Linux (“error: implicit declaration of function 'ptsname'”), and improved the error checks in ‘tty_spawn’.

I’m not sure whether the library should be given a more distinct name. There’s an [existing ‘libtty’](https://packages.debian.org/bookworm/libtty1).

Rinzwind committed:
Merged commit 32b2be55c15f2e97 (latest ‘pharo10’).

Rinzwind committed:
Changed error checking macros used in the child process in ‘tty_spawn’ to use ‘_exit’ instead of ‘exit’, and ‘dprintf’ instead of ‘printf’, to avoid flushing the open streams copied, buffers included, from the parent process.

Rinzwind commented:
Can this be merged? I had to make one more change (commit 6fa0c9e28884ae95) but I think it’s ready now. I wrote some tests for which I will open a Pharo pull request once ‘libtty’ is available in the VM.
